File: java/6.java
public class Solution {
    public String convert(String s, int numRows) {
        char[] res = new char[s.length()];
        int index = 0;
        int n = 1;
        int count = 0;
        int padding = 2 * (numRows - 1); // if n(numRows) == 4, padding = 6;
        if(numRows == 1) return s;
        for(int i = 0; i < res.length; i++){
            res[i] = s.charAt(index);
            if(n!=1 && n!= numRows){
                int p = 2 * (numRows - n);
                if(i+1 < res.length && index+p < s.length())
                    res[++i] = s.charAt(index+p);
            }
            index = index + padding;
            if(index >= s.length()){
                index = ++count;
                n++;
            }
        }

        return new String(res);
    }
}

File: C++/6.c
char* convert(char* s, int numRows) {
	if(numRows==1)
		return s;
	else if(numRows==2)
	{
		char* s1=calloc(strlen(s)+1,sizeof(char));
		int index=0;
		for(int i=0;i<strlen(s);i+=2,index++)
			s1[index]=s[i];
		for(int i=1;i<strlen(s);i+=2,index++)
			s1[index]=s[i];
		return s1;
	}
	else
	{
		char* s2=calloc(strlen(s)+1,sizeof(char));
		int in=0;
		int pass=(numRows-1)*2;
		int save=strlen(s)%pass-1;//余数
		for(int i=0;i<strlen(s);i+=pass,in++)
			s2[in]=s[i];
		for(int i=1;i<numRows-1;i++)
		{
			for(int n=0;n<strlen(s)/pass;n++,in+=2)
			{
				s2[in]=s[n*pass+i];
				s2[in+1]=s[n*pass+pass-i];
			}
			if(save>=i)
			{
				s2[in]=s[strlen(s)/pass*pass+i];
				in++;
			}
			if(save>=(pass-i))
			{
				s2[in]=s[strlen(s)/pass*pass+pass-i];
				in++;
			}
		}
		for(int i=pass/2;i<strlen(s);i+=pass,in++)
			s2[in]=s[i];
		return s2;
	}
	
	
}